File: leetcode/unique-paths-iii/unique-paths-iii.py
# source: https://leetcode.com/problems/unique-paths-iii/
# using pythons solution as prototype

from typing import List


class Solution:
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        count = 0
        rows = len(grid)
        cols = len(grid[0])

        def find_start():
            for i in range(len(grid)):
                for j in range(len(grid[0])):
                    if grid[i][j] == 1:
                        return i, j

        def count_zeros():
            zeros = 0
            for i in range(len(grid)):
                for j in range(len(grid[0])):
                    if grid[i][j] == 0:
                        zeros += 1
            return zeros

        istart, jstart = find_start()

        def dfs(i, j, z):
            if grid[i][j] == 2:
                if z != -1:
                    return
                nonlocal count
                count += 1
                return

            grid[i][j] = 1

            if i + 1 < rows and (grid[i + 1][j] == 0 or grid[i + 1][j] == 2):
                dfs(i + 1, j, z - 1)

            if j + 1 < cols and (grid[i][j + 1] == 0 or grid[i][j + 1] == 2):
                dfs(i, j + 1, z - 1)

            if i - 1 > -1 and (grid[i - 1][j] == 0 or grid[i - 1][j] == 2):
                dfs(i - 1, j, z - 1)

            if j - 1 > -1 and (grid[i][j - 1] == 0 or grid[i][j - 1] == 2):
                dfs(i, j - 1, z - 1)

            grid[i][j] = 0

        dfs(istart, jstart, count_zeros())
        return count




if __name__ == '__main__':
    sol = Solution()
    print("Expected: 2, Output: ", sol.uniquePathsIII([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 2, -1]]))
    print("Expected: 4, Output: ", sol.uniquePathsIII([[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 2]]))
    print("Expected: 0, Output: ", sol.uniquePathsIII([[0, 1], [2, 0]]))

File: leetcode/unique-paths-iii/unique-paths-iii.go
package main

import "fmt"

// source: https://leetcode.com/problems/unique-paths-iii/

var res = 0

func dfs(grid [][]int, i, j, z int) {
	if grid[i][j] == 2 {
		if z != -1 {
			return
		}
		res++
		return
	}

	grid[i][j] = 1

	if i+1 < len(grid) && (grid[i+1][j] == 0 || grid[i+1][j] == 2) {
		dfs(grid, i+1, j, z-1)
	}

	if j+1 < len(grid[0]) && (grid[i][j+1] == 0 || grid[i][j+1] == 2) {
		dfs(grid, i, j+1, z-1)
	}

	if i-1 > -1 && (grid[i-1][j] == 0 || grid[i-1][j] == 2) {
		dfs(grid, i-1, j, z-1)

	}

	if j-1 > -1 && (grid[i][j-1] == 0 || grid[i][j-1] == 2) {
		dfs(grid, i, j-1, z-1)
	}
	grid[i][j] = 0
}

func uniquePathsIII(grid [][]int) int {
	var iStart, jStart, zeros int
	for i := range grid {
		for j := range grid[i] {
			if grid[i][j] == 0 {
				zeros++
			} else if grid[i][j] == 1 {
				iStart, jStart = i, j
			}
		}
	}

	res = 0
	dfs(grid, iStart, jStart, zeros)
	return res
}

func main() {
	// Example 1
	var grid1 = [][]int{{1, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 2, -1}}
	fmt.Println("Expected: 2	Output: ", uniquePathsIII(grid1))

	// Example 2
	var grid2 = [][]int{{1, 0, 0, 0}, {0, 0, 0, 0}, {0, 0, 0, 2}}
	fmt.Println("Expected: 4	Output: ", uniquePathsIII(grid2))

	// Example 3
	var grid3 = [][]int{{0, 1}, {2, 0}}
	fmt.Println("Expected: 0	Output: ", uniquePathsIII(grid3))

}
